server/task_check_scheduler.go

```diff
@@ -177,8 +177,77 @@ func (s *TaskCheckScheduler) Register(taskType string, executor TaskCheckExecuto
 	s.executors[taskType] = executor
 }
 
+// shouldScheduleTimingTaskCheck will return whether should we schedule a timing task check for current task
+// would return true when:
+// 1. user specified the notBeforeTs field, hence a timing gate is added
+// 2. there is no running timing task check
+// 3. the notBeforeTs has altered since last check (either creation or patch)
+func (s *TaskCheckScheduler) shouldScheduleTimingTaskCheck(ctx context.Context, task *api.Task) (bool, error) {
+	if task.NotBeforeTs == 0 {
+		return false, nil
+	}
+
+	statusList := []api.TaskCheckRunStatus{api.TaskCheckRunDone, api.TaskCheckRunFailed, api.TaskCheckRunRunning}
+	taskCheckType := api.TaskCheckGeneralEarliestAllowedTime
+	taskCheckRunFind := &api.TaskCheckRunFind{
+		TaskID:     &task.ID,
+		Type:       &taskCheckType,
+		StatusList: &statusList,
+		Latest:     true,
+	}
+	taskCheckRunList, err := s.server.TaskCheckRunService.FindTaskCheckRunList(ctx, taskCheckRunFind)
+	if err != nil {
+		return false, err
+	}
+
+	if len(taskCheckRunList) == 0 {
+		return true, nil
+	}
+
+	if taskCheckRunList[0].Status == api.TaskCheckRunRunning {
+		return false, nil
+	}
+
+	taskCheckPayload := &api.TaskCheckEarliestAllowedTimePayload{}
+	if err := json.Unmarshal([]byte(taskCheckRunList[0].Payload), taskCheckPayload); err != nil {
+		return false, err
+	}
+
+	if taskCheckPayload.NotBeforeTs != task.NotBeforeTs {
+		return true, nil
+	}
+
+	return false, nil
+}
+
 // ScheduleCheckIfNeeded schedules a check if needed.
 func (s *TaskCheckScheduler) ScheduleCheckIfNeeded(ctx context.Context, task *api.Task, creatorID int, skipIfAlreadyTerminated bool) (*api.Task, error) {
+	// the following block is for timing task check
+	{
+		flag, err := s.shouldScheduleTimingTaskCheck(ctx, task)
+		if err != nil {
+			return nil, err
+		}
+
+		if flag {
+			taskCheckPayload, err := json.Marshal(task.NotBeforeTs)
+			if err != nil {
+				return nil, err
+			}
+			_, err = s.server.TaskCheckRunService.CreateTaskCheckRunIfNeeded(ctx, &api.TaskCheckRunCreate{
+				CreatorID:               creatorID,
+				Payload:                 string(taskCheckPayload),
+				TaskID:                  task.ID,
+				Type:                    api.TaskCheckGeneralEarliestAllowedTime,
+				SkipIfAlreadyTerminated: false,
```

Comment:
I think we should still somehow respect skipIfAlreadyTerminated, we may move the logic taskCheckPayload.NotBeforeTs != task.NotBeforeT here to determine